README.md
# postcss-auto2remvw

[![test](https://github.com/zqcccc/postcss-auto2remvw/actions/workflows/codecov.yml/badge.svg)](https://github.com/zqcccc/postcss-auto2remvw)
[![Codecov Coverage](https://img.shields.io/codecov/c/github/zqcccc/postcss-auto2remvw/main.svg)](https://codecov.io/gh/zqcccc/postcss-auto2remvw/)
<a href="https://npmcharts.com/compare/postcss-auto2remvw?minimal=true"><img src="https://img.shields.io/npm/dm/postcss-auto2remvw.svg?sanitize=true" alt="Downloads"></a>
<a href="https://www.npmjs.com/package/postcss-auto2remvw"><img src="https://img.shields.io/npm/v/postcss-auto2remvw.svg?sanitize=true" alt="Version"></a>
<a href="https://www.npmjs.com/package/postcss-auto2remvw"><img src="https://img.shields.io/npm/l/postcss-auto2remvw.svg?sanitize=true" alt="License"></a>

[PostCSS](https://github.com/postcss/postcss) plugin auto change px to rem & vw.

## Intro

The rem layout is affected by the root font-size, if Android turns on the font-size enlargement mode, it will cause the whole layout to crash.

The viewport scale under Android is forced to 1, some Canvas drawing content is slightly blurred. Some Android devices have bug in changing scale, too many devices are difficult to locate the problem.

This plugin will increase the size of css。

```css
.foo {
  height: 100px;
}
```

```css
.foo {
  height: 0.133333rem;
  height: 13.333333vw;
}
```

## Usage

**Step 1:** Install plugin:

```sh
npm install --save-dev postcss postcss-auto2remvw
```

**Step 2:** Check you project for existed PostCSS config: `postcss.config.js`
in the project root, `"postcss"` section in `package.json`
or `postcss` in bundle config.

If you do not use PostCSS, add it according to [official docs]
and set this plugin in settings.

**Step 3:** Add the plugin to plugins list:

```diff
module.exports = {
  plugins: [
+   require('postcss-auto2remvw')(config),
    require('autoprefixer')
  ]
}
```

## Config

| Name              | Type                               | Default                 | Description                                                  |
| ----------------- | ---------------------------------- | ----------------------- | ------------------------------------------------------------ |
| *baseSize*        | `{ rem: number, vw: number }`      | `{ rem: 750, vw: 7.5 }` | Base unit of the design draft for comparison with px ( If you change this value you must have both rem and vw ) |
| *precision*       | `int`                         | `6`                     | Calculation accuracy                                         |
| *forceRemProps*   | `string[]`               | `["font", "font-size"]` | Force only rem units to be used for properties               |
| *keepRuleComment* | `string`                           | `"no-RemVw-next-line"`  | Disable next line to use auto2remvw, must use like `/* no-RemVw-next-line */` |
| *keepFileComment* | `string`                           | `"auto2remvw-disable"`  | Disable whole file to use auto2remvw ( must be the first line of the file ) |
| *customConfigs*   | `{ test: RegExp, config: Config }[]` | `undefined`             | Specify part of the file to use other configurations         |




[postcss official docs]( https://github.com/postcss/postcss#usage)


CHANGELOG.md
# Change Log

## 0.0.2

customConfig property reg to test

## 0.0.1

- px to rem & vw

This project adheres to [Semantic Versioning](http://semver.org/).


test/index.test.js
const postcss = require("postcss")
const { join } = require("path")

const plugin = require("..")

async function easeRun(
  input,
  output,
  opts ,
  { path = "normal", warningNum = 0, warningText = "" } = {}
) {
  let result = await postcss([plugin(opts)]).process(input, {
    from: join(__dirname, path),
  })
  expect(result.css).toEqual(output)
  const warnings = result.warnings()
  expect(warnings).toHaveLength(warningNum)
  if (warningNum) {
    expect(warnings[0].text).toEqual(warningText)
  }
}

it("default force rem", async () => {
  await easeRun(
    `
  .demo {
    font-size: 14px;
  }
  `,
    `
  .demo {
    font-size: 0.018667rem;
  }
  `,
  )
})

it("config force rem", async () => {
  await easeRun(
    `
  .demo {
    font-size: 14px;
  }
  `,
    `
  .demo {
    font-size: 0.018667rem;
    font-size: 1.866667vw;
  }
  `,
    {
      forceRemProps: [],
    }
  )
})

it("keyframes", async () => {
  await easeRun(
    `
  @keyframes jump {
    from {
      width: 2px;
    }
    to {
      width: 10px;
    }
  }
  `,
    `
  @keyframes jump {
    from {
      width: 0.002667rem;
      width: 0.266667vw;
    }
    to {
      width: 0.013333rem;
      width: 1.333333vw;
    }
  }
  `,
    {}
  )
})

it("keepRuleComment", async () => {
  await easeRun(
    `
  .demo {
    /* no-RemVw-next-line */
    width: 100px;
    height: 100px;
  }
  `,
    `
  .demo {
    /* no-RemVw-next-line */
    width: 100px;
    height: 0.133333rem;
    height: 13.333333vw;
  }
  `,
    {}
  )
})

it("keepFileComment", async () => {
  await easeRun(
    `
  /* auto2remvw-disable */
  .demo {
    width: 100px;
    height: 100px;
  }
  `,
    `
  /* auto2remvw-disable */
  .demo {
    width: 100px;
    height: 100px;
  }
  `,
    {}
  )
})

it("media", async () => {
  await easeRun(
    `
  @media only screen and (device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2) {
    .bottomWrapper {
      height: 144px;
      padding-bottom: 66px;
    }
  }
  `,
    `
  @media only screen and (device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2) {
    .bottomWrapper {
      height: 0.192rem;
      height: 19.2vw;
      padding-bottom: 0.088rem;
      padding-bottom: 8.8vw;
    }
  }
  `,
    {}
  )
})

it("exclude file", async () => {
  await easeRun(
    `
  .demo {
    font-size: 14px;
  }
  `,
    `
  .demo {
    font-size: 14px;
  }
  `,
    { exclude: /normal/g }
  )
})
it("exclude file is not exist", async () => {
  await easeRun(
    `
  .demo {
    font-size: 14px;
  }
  `,
    `
  .demo {
    font-size: 0.018667rem;
  }
  `,
    { exclude: /normal1/g }
  )
})
it("config.exclude is not reg", async () => {
  await easeRun(
    `
  .demo {
    font-size: 14px;
  }
  `,
    `
  .demo {
    font-size: 0.018667rem;
  }
  `,
    { exclude: "test" },
    {
      warningNum: 1,
      warningText: "exclude should be RegExp",
    }
  )
})

it("no baseSize", async () => {
  await easeRun(
    `.demo { font-size: 14px; }`,
    `.demo { font-size: 14px; }`,
    { baseSize: undefined },
    {
      warningNum: 1,
      warningText: "baseSize should be set",
    }
  )
})


test/customeConfig.test.js
const postcss = require('postcss')
const { join } = require('path')

const plugin = require('..')

it("custom config's test is not a regexp", async () => {
  let result = await postcss([
    plugin({
      customConfigs: [
        {
          test: "test",
        },
      ],
    }),
  ]).process(`.demo {}`, { from: undefined })
  expect(result.css).toEqual(`.demo {}`)
  expect(result.warnings()).toHaveLength(1)
})

it('different path use different config', async () => {
  let p = postcss([plugin({
    customConfigs: [
      {
        test: new RegExp('tiny-app/custom-path', 'g'),
        config: {
          baseSize: {
            rem: 375, vw: 3.75
          }
        }
      }
    ]
  })])
  async function run (input, output, path = 'normal', warningNum = 0) {
    let result = await p.process(input, { from: join(__dirname, path) })
    expect(result.css).toEqual(output)
    expect(result.warnings()).toHaveLength(warningNum)
  }
  await run(`
  .demo {
    font-size: 14px;
  }
  `, `
  .demo {
    font-size: 0.018667rem;
  }
  `)
  await run(`
  .demo {
    font-size: 14px;
  }
  `, `
  .demo {
    font-size: 0.037333rem;
  }
  `, 'tiny-app/custom-path/index')
  await run(`
  .demo {
    font-size: 14px;
  }
  `, `
  .demo {
    font-size: 0.018667rem;
  }
  `)
})


index.js
const { isRegExp, toFixed } = require("./util")

const defaultConfig = {
  baseSize: { rem: 750, vw: 7.5 },
  precision: 6,
  forceRemProps: ["font", "font-size"],
  keepRuleComment: "no-RemVw-next-line",
  keepFileComment: "auto2remvw-disable",
}
const pxGlobalRegExp = /\b(\d+(\.\d+)?)px\b/g

let cacheGlobalConfig
/**
 * @type {import('postcss').PluginCreator}
 */
module.exports = (options = {}) => {
  let config = Object.assign({}, defaultConfig, options)

  cacheGlobalConfig = config

  let isExcludeFile = false

  const getCalcValue = (targetUnit, sourceValue) => {
    const baseSize = config.baseSize[targetUnit]
    // if (!baseSize) return sourceValue

    return sourceValue.replace(pxGlobalRegExp, ($0, $1) => {
      return `${toFixed($1 / baseSize, config.precision)}${targetUnit}`
    })
  }
  return {
    postcssPlugin: "postcss-auto2remvw",

    Once(root, { result }) {
      const filePath = root.source.input.file
      if (!config.baseSize) {
        result.warn("baseSize should be set")
        return
      }
      if (config.exclude) {
        if (!isRegExp(config.exclude)) {
          result.warn("exclude should be RegExp")
          return
        }
        if (filePath.match(config.exclude) !== null) {
          isExcludeFile = true
        }
      }
      if (Array.isArray(config.customConfigs)) {
        config.customConfigs.forEach((schema) => {
          if (!isRegExp(schema.test)) {
            result.warn("custom config's reg should be a RegExp")
            return
          }
          if (schema.test.test(filePath)) {
            config = Object.assign({}, config, schema.config)
          }
        })
      }
    },

    Declaration(decl) {
      if (isExcludeFile || !config.baseSize) return
      const pre = decl.prev()
      if (
        pre &&
        pre.type === "comment" &&
        pre.text.trim() === config.keepRuleComment
      ) {
        return
      }
      const px = decl.value
      const rem = getCalcValue("rem", px)
      if (px === rem) return
      if (config.forceRemProps.indexOf(decl.prop) > -1) {
        decl.value = rem
        return
      }
      decl.value = getCalcValue("vw", px)
      decl.cloneBefore({
        prop: decl.prop,
        value: rem,
      })
    },

    Comment(comment) {
      if (comment.text.trim() === config.keepFileComment) {
        isExcludeFile = true
      }
    },
    OnceExit() {
      config = cacheGlobalConfig
    },
  }
}

module.exports.postcss = true
